src/sink.rs

```diff
@@ -180,33 +182,64 @@ where
         request: Request<Streaming<proto::SinkRequest>>,
     ) -> Result<tonic::Response<proto::SinkResponse>, Status> {
         let mut stream = request.into_inner();
-
+        let sink_handle = self.handler.clone();
+        let cancellation_token = self.cancellation_token.clone();
+        let shutdown_tx = self.shutdown_tx.clone();
         // TODO: what should be the idle buffer size?
         let (tx, rx) = mpsc::channel::<SinkRequest>(1);
 
-        // call the user's sink handle
-        let sink_handle = self.handler.sink(rx);
+        let writer_cln_token = cancellation_token.clone();
 
-        // write to the user-defined channel
+        // spawn a task to read messages from the stream and send them to the user's sink handle
         tokio::spawn(async move {
```

Comment:
get the handle and check explicitly for status or abort of cancellation token has been invoked.